File: cpp/test/TestParselet.cpp

#include "Tokenizer.h"
#include "ParseletRegistration.h"
#include "Parselet.h"
#include "ParserSession.h"

#include "gtest/gtest.h"

class ParserSession;

using ParserSessionPtr = ParserSession *;


class ParseletTest : public ::testing::Test {
protected:
    
    static ParserSessionPtr session;
    
    static void SetUpTestSuite() {
        
        session = new ParserSession();
    }
    
    static void TearDownTestSuite() {
        
        delete session;
    }
    
    void SetUp() override {
        
    }
    
    void TearDown() override {
        
        session->deinit();
    }

};

ParserSessionPtr ParseletTest::session;


TEST_F(ParseletTest, Bug1) {
    
    auto strIn = std::string("a /: b := c");

    auto str = reinterpret_cast<const unsigned char *>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(BufferAndLength(str, strIn.size()), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    (prefixParselets[Tok.Tok.value()]->parsePrefix())(session, prefixParselets[Tok.Tok.value()], Tok);

    auto& N = Parser_topNode(session);

    EXPECT_TRUE(std::holds_alternative<NodePtr>(N));
    
    auto& P = std::get<NodePtr>(N);
    
    EXPECT_TRUE(dynamic_cast<TernaryNode *>(P));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

//
// This used to assert
//
TEST_F(ParseletTest, Bug2) {
//
    auto strIn = std::string("a<b ");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(BufferAndLength(str, strIn.size()), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    (prefixParselets[Tok.Tok.value()]->parsePrefix())(session, prefixParselets[Tok.Tok.value()], Tok);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

//
// This used to assert
//
TEST_F(ParseletTest, Bug3) {
    
    auto strIn = std::string("a\\[Integral]b\\[Integral]c ");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(BufferAndLength(str, strIn.size()), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    (prefixParselets[Tok.Tok.value()]->parsePrefix())(session, prefixParselets[Tok.Tok.value()], Tok);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

//
// This used to assert
//
TEST_F(ParseletTest, Bug4) {
    
    auto strIn = std::string("\\[RawLeftBrace]*\\[RawRightBrace]");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(BufferAndLength(str, strIn.size()), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    (prefixParselets[Tok.Tok.value()]->parsePrefix())(session, prefixParselets[Tok.Tok.value()], Tok);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

File: crates/wolfram-parser/src/tests/test_parselet.rs
use crate::{cst::Cst, ParseOptions, ParserSession};


#[test]
fn ParseletTest_Bug1() {
    let strIn = "a /: b := c";

    let mut session =
        ParserSession::new(strIn.as_bytes(), &ParseOptions::default());

    let tok = session.tokenizer.peek_token();

    session.parse_prefix(tok);

    let P: &mut Cst<_> = session.builder.top_node();

    assert!(matches!(P, Cst::Ternary(_)));

    assert_eq!(session.tokenizer.non_fatal_issues.len(), 0);
    assert_eq!(session.tokenizer.fatal_issues.len(), 0);
}

//
// This used to assert
//
#[test]
fn ParseletTest_Bug2() {
    //
    let strIn = "a<b ";

    let mut session =
        ParserSession::new(strIn.as_bytes(), &ParseOptions::default());

    let tok = session.tokenizer.peek_token();

    session.parse_prefix(tok);

    assert_eq!(session.tokenizer.non_fatal_issues.len(), 0);
    assert_eq!(session.tokenizer.fatal_issues.len(), 0);
}

//
// This used to assert
//
#[test]
fn ParseletTest_Bug3() {
    let strIn = "a\\[Integral]b\\[Integral]c ";

    let mut session =
        ParserSession::new(strIn.as_bytes(), &ParseOptions::default());

    let tok = session.tokenizer.peek_token();

    session.parse_prefix(tok);

    assert_eq!(session.tokenizer.non_fatal_issues.len(), 0);
    assert_eq!(session.tokenizer.fatal_issues.len(), 0);
}

//
// This used to assert
//
#[test]
fn ParseletTest_Bug4() {
    let strIn = "\\[RawLeftBrace]*\\[RawRightBrace]";

    let mut session =
        ParserSession::new(strIn.as_bytes(), &ParseOptions::default());

    let tok = session.tokenizer.peek_token();

    session.parse_prefix(tok);

    assert_eq!(session.tokenizer.non_fatal_issues.len(), 0);
    assert_eq!(session.tokenizer.fatal_issues.len(), 0);
}
